pkg/controller/sparkapplication/controller.go

```diff
@@ -615,61 +613,57 @@ func (c *Controller) submitSparkApplication(app *v1alpha1.SparkApplication) *v1a
-
-		// Update AppStatus to submitted.
-		app.Status = v1alpha1.SparkApplicationStatus{
-			AppState: v1alpha1.ApplicationState{
-				State: v1alpha1.SubmittedState,
-			},
-			SubmissionAttempts: app.Status.SubmissionAttempts + 1,
-			ExecutionAttempts:  app.Status.ExecutionAttempts + 1,
-			SubmissionTime:     metav1.Now(),
-		}
-		c.recordSparkApplicationEvent(app)
-
-		// Add driver as a finalizer to prevent SparkApplication deletion till driver is deleted.
-		if app.ObjectMeta.Finalizers == nil {
-			app.ObjectMeta.Finalizers = []string{sparkDriverRole}
 		} else {
-			app.ObjectMeta.Finalizers = append(app.ObjectMeta.Finalizers, sparkDriverRole)
+			glog.Errorf("failed to run spark-submit for SparkApplication %s in namespace: %s. Error: %v", submission.name,
+				submission.namespace, errorMsg)
+			app.Status = v1alpha1.SparkApplicationStatus{
+				AppState: v1alpha1.ApplicationState{
+					State:        v1alpha1.FailedSubmissionState,
+					ErrorMessage: errorMsg,
+				},
+				SubmissionAttempts:        app.Status.SubmissionAttempts + 1,
+				LastSubmissionAttemptTime: metav1.Now(),
+			}
+			c.recordSparkApplicationEvent(app)
+			return app
 		}
+	}
 
-		if app.Spec.Monitoring != nil && app.Spec.Monitoring.Prometheus != nil {
-			// configPrometheusMonitoring may update app.Spec.
-			configPrometheusMonitoring(app, c.kubeClient)
-		}
+	// Submission Successful or Driver Pod Already exists.
+	glog.Infof("spark-submit completed for SparkApplication %s in namespace %s", submission.name, submission.namespace)
+	// Update AppStatus to submitted.
+	app.Status = v1alpha1.SparkApplicationStatus{
+		AppState: v1alpha1.ApplicationState{
+			State: v1alpha1.SubmittedState,
+		},
+		SubmissionAttempts:        app.Status.SubmissionAttempts + 1,
+		ExecutionAttempts:         app.Status.ExecutionAttempts + 1,
+		LastSubmissionAttemptTime: metav1.Now(),
+	}
+	c.recordSparkApplicationEvent(app)
 
-		// Create Spark UI Service.
-		service, err := createSparkUIService(app, c.kubeClient)
-		if err != nil {
-			glog.Errorf("Failed to create UI service for SparkApplication %s/%s. Error: %v", app.Namespace, app.Name, err)
-		} else {
-			app.Status.DriverInfo.WebUIServiceName = service.serviceName
-			app.Status.DriverInfo.WebUIPort = service.nodePort
-			// Create UI Ingress if ingress-format is set.
-			if c.ingressUrlFormat != "" {
-				ingress, err := createSparkUIIngress(app, *service, c.ingressUrlFormat, c.kubeClient)
-				if err != nil {
-					glog.Errorf("Failed to create UI Ingress for SparkApplication %s/%s. Error: %v", app.Namespace, app.Name, err)
-				} else {
-					app.Status.DriverInfo.WebUIIngressAddress = ingress.ingressUrl
-					app.Status.DriverInfo.WebUIIngressName = ingress.ingressName
-				}
+	// Add driver as a finalizer to prevent SparkApplication deletion till driver is deleted.
+	app = addFinalizer(app, sparkDriverRole)
+	if app.Spec.Monitoring != nil && app.Spec.Monitoring.Prometheus != nil {
```

Comment:
Hmm, Prometheus monitoring configuration should be done before submitting the application.

Reply:
done, I had to clone the App since we modify the Spec which will cause a onUpdate callback and a clean-up/re-run of the Application. Not sure if this is already an issue in master.

Reply:
In the current master, only the status section is updated to the API server, so this is not an issue. But this is no longer the case with the changes in this PR as you add a finalizer to the mix.